internal/jobobject/queue.go

```diff
@@ -0,0 +1,96 @@
+package jobobject
+
+import (
+	"errors"
+	"sync"
+)
+
+var (
+	errQueueClosed = errors.New("the queue is closed for reading and writing")
+	errQueueEmpty  = errors.New("the queue is empty")
+)
+
+// messageQueue represents a threadsafe message queue to be used to retrieve or
+// write notifications to. This will be used to hold and retrieve messages for job
+// objects and doesn't have much use in other packages. There should only be one reader
+// and one writer.
+type messageQueue struct {
+	m        sync.Mutex
+	messages []interface{}
+	ready    chan struct{} // Channel to signal that a message is available to read.
+	closed   bool
+}
+
+func newMessageQueue() *messageQueue {
+	return &messageQueue{
+		messages: []interface{}{},
+		ready:    make(chan struct{}),
+	}
+}
+
+func (mq *messageQueue) write(msg interface{}) error {
+	mq.m.Lock()
+	defer mq.m.Unlock()
+
+	if mq.closed {
+		return errQueueClosed
+	}
+	length := len(mq.messages)
+	mq.messages = append(mq.messages, msg)
+	// If the queue was empty, we need to signal that there's now a value ready to read.
+	if length == 0 {
+		// This handles the very first write case. If this is the first write, we could end
+		// up in a spot where we'd signal mq.ready but no one is waiting for a value so we'll block
+		// forever. This way we can add to the queue, so any call to read after will just return the
+		// value and any call to readOrWait should be listening for a new value and we'll hit the
+		// `mq.ready <- struct{}{}` case.
+		select {
+		case mq.ready <- struct{}{}:
+		default:
+		}
+	}
+	return nil
+}
+
+// read will read a value from the queue if available, otherwise return an error.
+func (mq *messageQueue) read() (interface{}, error) {
+	mq.m.Lock()
+	defer mq.m.Unlock()
+	if mq.closed {
+		return nil, errQueueClosed
+	}
+	if len(mq.messages) == 0 {
+		return nil, errQueueEmpty
+	}
+	val := mq.messages[0]
+	mq.messages[0] = nil
+	mq.messages = mq.messages[1:]
```

Comment:
I'm pretty sure this will panic if the length of the queue was initially 1

Reply:
If the length is 1 this works fine, you can test here. https://play.golang.org/p/177qJbPuACa Unless you're referring to something else.